### PR TITLE
Fix build for macOS Sierra + LLVM 3.9.

### LINKED ISSUE #19762: build failure on macOS
fresh clone of master branch at 19f81ac
```
[ 75%] Building CXX object lib/ExecutionEngine/Orc/CMakeFiles/LLVMOrcJIT.dir/OrcMCJITReplacement.cpp.o
[ 75%] Building CXX object lib/ExecutionEngine/Orc/CMakeFiles/LLVMOrcJIT.dir/OrcRemoteTargetRPCAPI.cpp.o
In file included from /Users/sachs/src/julia-master/deps/srccache/llvm-3.9.1/lib/ExecutionEngine/Orc/OrcRemoteTargetRPCAPI.cpp:10:
In file included from /Users/sachs/src/julia-master/deps/srccache/llvm-3.9.1/include/llvm/ExecutionEngine/Orc/OrcRemoteTargetRPCAPI.h:21:
In file included from /Users/sachs/src/julia-master/deps/srccache/llvm-3.9.1/include/llvm/ExecutionEngine/Orc/RPCUtils.h:35:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/future:1447:23: error:
      'future_error' is unavailable: introduced in macOS 10.8
                      future_error(make_error_code(future_errc::broken_promise))
                      ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/future:502:63: note:
      'future_error' has been explicitly marked unavailable here
class _LIBCPP_EXCEPTION_ABI _LIBCPP_AVAILABILITY_FUTURE_ERROR future_error
                                                              ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/future:1621:23: error:
      'future_error' is unavailable: introduced in macOS 10.8
                      future_error(make_error_code(future_errc::broken_promise))
                      ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/future:502:63: note:
      'future_error' has been explicitly marked unavailable here
class _LIBCPP_EXCEPTION_ABI _LIBCPP_AVAILABILITY_FUTURE_ERROR future_error
                                                              ^
2 errors generated.
make[4]: *** [lib/ExecutionEngine/Orc/CMakeFiles/LLVMOrcJIT.dir/OrcRemoteTargetRPCAPI.cpp.o] Error 1
make[3]: *** [lib/ExecutionEngine/Orc/CMakeFiles/LLVMOrcJIT.dir/all] Error 2
make[2]: *** [all] Error 2
make[1]: *** [scratch/llvm-3.9.1/build_Release/build-compiled] Error 2
make: *** [julia-deps] Error 2
```

### PR DESCRIPTION
**Pull request by andreasnoack**
According to [this comment](https://github.com/JuliaLang/julia/blob/3b5c9755b7e2b8f7b4e202ff2d4d3dc07d6decca/Make.inc#L76) this change shouldn't be a problem.

(edit by tkelman - closes https://github.com/JuliaLang/julia/issues/19762)

### CODE REVIEW

**Comment by tkelman** [comment]
We'd need to update all places that list 10.7 as the minimum supported version.

**Comment by andreasnoack** [comment]
I'm not sure that this changes the minimum required version. This PR only changes the minimum requirement when `USE_LIBCPP=1` and that is only the default for OS X >= 10.9 .

**Comment by tkelman** [comment]
If the binaries get built on a newer system than that, then this will change the minimum supported version.

**Comment by andreasnoack** [comment]
So https://status.julialang.org/download/osx10.7+ should be changed. Are there other places?

**Comment by tkelman** [comment]
The downloads page, readme, documentation, anywhere else we refer to 10.7

**Comment by andreasnoack** [comment]
I've changed the places that matched "10.7" but I didn't see any matches in the documentation and the readme is about building Julia which is still possible on 10.7 as mentioned. Is the download link on the download page automatically generated or should I open PR for the homepage? Please be specific if you want other places to be changed.

File: contrib/prepare_release.sh [hunk 1]
```diff
@@ -46,9 +46,9 @@ cp julia-$version-linux-arm.tar.gz julia-$majmin-latest-linux-arm.tar.gz
 curl -L -o julia-$version-linux-ppc64le.tar.gz \
   $julianightlies/linux/ppc64le/$majmin/julia-$majminpatch-$shashort-linuxppc64.tar.gz
 cp julia-$version-linux-ppc64le.tar.gz julia-$majmin-latest-linux-ppc64le.tar.gz
-curl -L -o "julia-$version-osx10.7 .dmg" \
+curl -L -o "julia-$version-osx10.8 .dmg" \
   $julianightlies/osx/x64/$majmin/julia-$majminpatch-$shashort-osx.dmg
-cp "julia-$version-osx10.7 .dmg" "julia-$majmin-latest-osx10.7 .dmg"
+cp "julia-$version-osx10.8 .dmg" "julia-$majmin-latest-osx10.8 .dmg"
```
Comment (tkelman) [inline contrib/prepare_release.sh]: this isn't correct unless the buildbots get changed

**Comment by tkelman** [comment]
The downloads page is not autogenerated, so yes check the website as well.

I don't know where they all are, we'll need to look and double-check to avoid leaving incorrect information. The buildbots determine the binary names in many cases.

**Comment by andreasnoack** [comment]
@tkelman Could you please recap the steps most convenient for merging of this. Should the buildbot setup change before or after this PR to minimize the problems/misleading info?

**Comment by tkelman** [comment]
This is somewhat coupled to https://github.com/JuliaLang/julia/pull/18554 which is proposing renaming the binary download file names, but isn't quite settled yet in terms of all the various places that will need to have links updated.

The changes should at least be prepared before this merges. That way once they're ready we can merge everything around the same time.

**Comment by tkelman** [comment]
> We'd need to update all places that list 10.7 as the minimum supported version.

@staticfloat how many places and filenames is this that you can think of? release filenames, links, buildbot info...

Alternately, could we stub out `future_error` so our LLVM 3.9 build maintains osx 10.7 compatibility to fix #19762?

**Comment by tkelman** [comment]
I'm going to push a commit to this branch that makes the default build 10.8+, but allow setting a makefile variable to modify that. The buildbots don't build Julia from 10.12, so we can maybe keep it at 10.7 there by setting the variable appropriately.

**Comment by staticfloat** [comment]
> @staticfloat how many places and filenames is this that you can think of? release filenames, links, buildbot info...

The only places I can think of are the actual naming of binaries, so that would include download URLs and the like.  The only reason we had the `10.7+` on the end of filenames was we used to have a `10.6` version of Julia as well for a while.  As you've noted, #18554 will drop this completely, so perhaps we don't have to worry about this too much.

**Comment by ViralBShah** [comment]
10.8 was released 4 years ago. Seems reasonable to make it the min supported OS.

**Comment by tkelman** [comment]
If necessary and if dropping 10.7 gets us anything in return, sure. it shouldn't actually be needed yet if building on 10.11 or earlier and setting this variable to 10.7, which we can do on the buildbots.

**Comment by stevengj** [comment]
(Meanwhile, people on the mailing list are still complaining about Windows XP.  Thank goodness no one is bugging us for MacOS 10.1 support.)

**Comment by tkelman** [comment]
it would be a very small patch to bring back xp support, at least part of the way. don't think you can say that about powerpc 32 bit macs.

**Pull request merged**